### PR TITLE
core: introduce UseRootFileSystemNamespace option

### PR DESCRIPTION
**Pull request by marcinpe**
Allow user to specify that unit has to run in init's files system
namespace. Previously it was possible, iff none of the file system
namespace related option was set (but documentation gave no such
warranty). I wanted to have a formal and documented option which
forces such behavior. This option is incompatible with: PrivateTmp=,
PrivateDevices=, ProtectSystem=, ProtectHome=, ReadOnlyPaths=,
InaccessiblePaths=, ReadWritePaths=, MountFlags=, meaning it
cannot be set to true, if any of the listed options has non default
value.

Why this is useful?
We like a lot setup in which ssh + few other of our daemons run
in the root namespace. Thanks to that the mounts done by our
daemons are visible to users who ssh to a server to debug issues.
They are also visible to chef and our monitoring tools.
Of course this can also be achieved by having the mounts marked as
shared and not sharing the namespace. However, shared mounts
has proven problematic to us over time (especially how easy it is
to create cycles).
In the current state of systemd we can make ssh + our daemons run
in the root namespace, by making sure they do not use any of the
file system namespace related options. The dark side of this
approach is, how easy it is to break it. We recently specified
MountFlags=private and all worked nicely, until we realized
ssh cannot see the mounts done by our daemon (and our tools depend
on it). What is worse the documentation never promises that units
will run in root namespace, it just happens to be so.
As such we want to have an explicit option which forces unit
to not clone file system namespace from systemd, but stay in
the original one.

### CODE REVIEW

**Comment by evverx** [comment]
@marcinpe , thanks!

Why not to use

```
nsenter --mount \
    --target $(systemctl show --value --property MainPID some-service) \
    some-check-script
```

?

**Comment by marcinpe** [comment]
Hey @evverx,
Thanks for quick reply!
What you are proposing is doable and we did consider it. There are few things we dislike about it:
- the command you proposed becomes a pre run step for every script. You keep trying to remember which namespace you are in. Namespaces are fantastic to separate process, but become really messy when you keep switching between them :-(
- users who log in to debug always have to remember to do it, and easily can get misguided if they forget. 

To give a solid example, when I want to see some interesting log from 3-4 machines I do:

```
for m in $(machine_list); do
  ssh root@${m} 'cat /var/log/.../some-log' >> big-log-file
done
```

With the nsenter this becomes much more complex: 

```
for m in $(machine_list); do
  ssh root@${m} 'nsenter --mount \ --target $(systemctl show --value --property MainPID some-service) cat /var/log/.../some-log' >> big-log-file
done
```

To get my arguments more concrete, let me explain what I work on. My team is proving container infrastructure (yeah yet another one :P ). My users are developers who run their binaries on their servers. To make things safe/prevent people from aiming at their feet we use all the cool isolation mechanisms that kernel provides these days. 
However, we really love the fact that these mechanisms are implementation details, ie. our users do not need to know/understand them. When they login with ssh, they just see the root namespace which has full view of the world. They can use all the tools that are installed on the hosts (but are not visible to their binaries which run isolated in very bare environment), but also they see the logs which are within the chroot of the binary.

The above are 'human' reasons. Now is the time for technical reason:
The environment (container) which we create for binaries is really tiny (for example has no gdb). To make debugging reasonable one needs to have a namespace that can see the mounts from the container, but also see the rest of the filesystem (with all the important tools). To get there we have a daemon who creates all the containers and their chroots. This guy both can see the container mounts and the rest of filesystem. He is the root node. If we run this daemon in his own mount namespace, there is a problem: the namespace will disappear on daemon restart. The containers keep running when daemon restarts (so that we can easily update, rollback or whatever without affecting containers). However on restart the namespace in which daemon run disappears... We could keep it around by some tricks of course, but then when daemon comes back we would have two root namespaces and we would not know which one to enter. 

In other words, this patch is not needed if we could create 'persistent' file system namespace and configure services to always be started in specific namespace. I thought this is a huge feature and maybe an overkill ;-)

**Comment by tixxdz** [comment]
I would say by default they should stay in the same namespaces if none of the namespace options is used, and without the need to add an extra option... probably a bug ? anyway would `JoinsNamespaceOf=` + other namespace related options solve your case ? where `JoinsNamespaceOf=` applies to the namespace you want to join, the daemon that does mounts.

**Comment by marcinpe** [comment]
JoinsNamespaceOf does not really solve it. We want to always join some stable namespace that never goes away. That said every service of ours can restart, the only stable namespace is the one belonging to init(systemd). If JoinsNamespaceOf allowed saying: "force running in init's namespace" we would be perfectly happy.
Of course there is the alternative of just documenting explicitly that if none of the namespace related options is used, unit will run in init's namespace. What I dislike about it, is that it is an 'implicit' option. When you read the unit configuration you have to know beforehand that the lack of any of 7 other options implies that you request to run in specific namespace.

**Comment by poettering** [comment]
I am not convinced that adding such a "negating" option, that disables a good chunk of the others would be a good idea. I mean, let's note that none of the namespacing options defaults to on, which means, unless your unit file _explicitly_ enabled any of these options your service will live in the host's main namespace. Hence, adding an option that allows you to _explicitly_ turn off the effect of options you _explicitly_ turned on in the same file sounds less than ideal to me?

Maybe this should really be dealt with as a documentation issue. For example, I am tempted to rearrange the systemd.exec man page, and place all security/sandboxing options that detach the mount namespace in a section of their own explaining the side-effect of using them in regards to detaching the services from mounts on the host. Would that be helpful?

Does that above make any sense?

**Comment by marcinpe** [comment]
I think that what you proposed is the right solution. I agree that having a negating option that disables 7 other options is very ugly. The reason I wanted it in the first place is to avoid accidental breakage: a good willing soul sees one of the cool features, adds it to the unit without being aware of the side effects and disaster is ready. Having a very clear documentation goes a long way.

**Comment by poettering** [comment]
I filed a new issue now in #4393, just make sure we don't forget to add this.
